Implement Prometheus metrics for LocalQueue

KPostOffice opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind feature
/kind api-change

#### What this PR does / why we need it:

Implementation of LQ metrics KEP

this replaces PR #3609 

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #1833 

#### Special notes for your reviewer:

I'm uncertain if I've updated all the metrics in the right places. I still need to write tests, but I figured I'd open the PR as I have it now in case anything is egregiously off. 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Addition of configuration that allows users to get prometheus metrics about LQ info, including the LQ status and the status of pending workloads

metrics added:
local_queue_pending_workloads
local_queue_quota_reserved_workloads_total
local_queue_quota_reserved_wait_time_seconds
local_queue_admitted_workloads_total
local_queue_admission_wait_time_seconds
local_queue_admission_checks_wait_time_seconds
local_queue_evicted_workloads_total
local_queue_reserving_active_workloads
local_queue_admitted_active_workloads
local_queue_status
local_queue_resource_reservation
local_queue_resource_usage
```

KPostOffice commented:
Currently I've implemented this with just a boolean feature gate. I was having trouble figuring out how to pass namespace selector details down to the `scheduler`, `cache`, and `queue` and then also act on those selector values. I didn't want to introduce client calls to them since I figured making network request would pretty severely hurt performance in those packages.

mimowo commented on pkg/metrics/metrics.go:
```diff
@@ -102,6 +108,17 @@ The label 'result' can have the following values:
+'status' can have the following values:
+- "active" means that the workloads are in the admission queue.
+- "inadmissible" means there was a failed admission attempt for these workloads and they won't be retried until cluster conditions, which could make this workload admissible, change`,
+		}, []string{"local_queue", "namespace", "status"},
```
I think this is acceptable, but let me consider other options:
1. "local_queue", "namespace" - as in the proposal
2. "name", "namespace"
3. "local_queue" - key "namespace/name"

I'm not in favor of (3) because maybe for some use-cases one wants to aggregate metrics by LQ name rather than full key. 

My only slight preference for (2.) is that it is less redundant. It is already clear from the metrics name that we are talking LQs. This is not the case for the `pending_workloads` metrics for CQs, so I think we don't need to follow the naming pattern for params strictly here. WDYT?

KPostOffice replied:
I'm happy with `2`

mimowo commented on pkg/cache/clusterqueue.go:
```diff
@@ -250,6 +250,11 @@ func (c *clusterQueue) updateQueueStatus() {
 		c.Status = status
 		metrics.ReportClusterQueueStatus(c.Name, c.Status)
+		if lqMetrics {
+			for _, lq := range c.localQueues {
```
This iteration might be adding unnecessary performance cost. What is the scenario that it needs calling here? Maybe we could move the call per LQ, when we update the specific LQ. PTAL.

KPostOffice replied:
The lq status is equal to the cq status. So when the cq status updates, all the cqs associated lqs should have their statuses updated as well

mimowo reviewed (commented):
Overall very nice to see this contribution, and I would like to include it in 0.10 if time allows. Left some comments about major things which draw my attention during initial pass. I would also like to see some integration tests - I think for most of the metrics we should be able to extend the tests were we check metrics for CQs. 

cc @tenzen-y @dgrove-oss @PBundyra

mimowo commented:
@KPostOffice in the release note, please list all the metrics and their shortened description / purpose.

mbobrovskyi commented:
/ok-to-test

tenzen-y commented:
/retitle Implement Prometheus metrics for LocalQueue

PBundyra commented on apis/config/v1beta1/configuration_types.go:
```diff
@@ -146,6 +146,9 @@ type ControllerMetrics struct {
 	EnableClusterQueueResources bool `json:"enableClusterQueueResources,omitempty"`
+
+	// +optional
+	EnableLocalQueueMetrics bool `json:"enableLocalQueueMetrics,omitempty"`
```
I very much agree, especially when it comes to passing parameters

KPostOffice replied:
There was a comment about increasing [cardinality](https://github.com/kubernetes-sigs/kueue/issues/1833#issuecomment-1994473006) and wanting to leave this behind a long term config field

mimowo replied:
I see, but in that case I would like to go via the KEP process. Pity the comment does not mention why cardinality is a problem - is it for usability (this could be solved by aggregation), or performance. Do you have some other references why cardinality might be a problem in k8s. 

I assume we don't have many more LQs than namepaces, which also let me check what we do in the core k8s. I see that we have metrics depending on Namespace, [example](https://github.com/kubernetes/kubernetes/blob/810e9e212ec5372d16b655f57b9231d8654a2179/pkg/registry/core/pod/rest/metrics.go#L38C4-L38C13). However, in this case we use explicitly CounterOpts.Namespace. Maybe we could also do it this way? PTAL. 

If you want this feature in 0.10 I think the only chance is a short KEP, don't change API, and guard it by Alpha feature gate (disabled by default). Then for second iteration of alpha investigate if we need the API switch.

KPostOffice replied:
`https://kubernetes.io/docs/reference/instrumentation/metrics/`

There's a few metrics here with a namespace label.

There's an existing [KEP](https://github.com/kubernetes-sigs/kueue/pull/2516) that I was having a bit of trouble implementing since it included the ability to use namespace/`local_queue` selectors for metric collection.

mimowo replied:
I would suggest to update the KEP and hide the mretics behind alpha feature gate. This will not impact users /customers and we don't commit to maintain the API. Then as graduation point for beta we re-evaluate both approaches

mimowo replied:
Sure, but we will not be able to do this for 0.10. Let's update the KEP and revisit the knobs for 0.11 before Beta.

tenzen-y replied:
I'm ok with v0.11 due to the tighter schedule for v0.10, and this feature is still alpha.

PBundyra commented on pkg/queue/cluster_queue.go:
Can we move changes applied to this file to `pkg/queue/local_queue.go`?

mimowo commented:
@KPostOffice I think this PR still lacks integration tests, given this is early Alpha I would be willing to go ahead based on manual testing, but would like to fill in the gap as soon as possible (possibly yet before the release). WDYT? cc @tenzen-y

KPostOffice commented:
I've added some integration tests to the PR, I've also changed the LQ status metric. It now is a direct reflection of the CRs active status. @mimowo Let me know if this is better than updating the status directly in the cache. It changes the values to `True`, `False`, `Unknown`. The only drawback with this is some lag to the internal status as well as losing the `Terminating` status but it avoids looping over all LQs every time the CQ's status is updated

mimowo commented on pkg/metrics/metrics.go:
```diff
@@ -269,29 +385,67 @@ func QuotaReservedWorkload(cqName kueue.ClusterQueueReference, waitTime time.Dur
+	}
+}
+
+func LQRefFromLocalQueueKey(wlKey string) LocalQueueReference {
```
```suggestion
func LQRefFromLocalQueueKey(lqKey string) LocalQueueReference {
```
nit

mimowo commented:
> . @mimowo Let me know if this is better than updating the status directly in the cache. It changes the values to True, False, Unknown. The only drawback with this is some lag to the internal status as well as losing the Terminating status but it avoids looping over all LQs every time the CQ's status is updated

I see, we can still experiment and collect feedback as the feature is alpha, so we can adjust later. For now, I prefer the option without iterating over the LQs in cache

mimowo commented:
I thought a little bit more about the knob and metrics cardinality:
1. we have the enable/disable knob for CQs, so I think this makes sense for LQs too: https://github.com/kubernetes-sigs/kueue/blob/e5e340da00732ccf61df4b3bbf944783c35e66a8/apis/config/v1beta1/configuration_types.go#L148 (by analogy)
2. the cardinality of namespace is ok based on the k8s metrics, `https://kubernetes.io/docs/reference/instrumentation/metrics/`
3. too much cardinality at some points become a problem surely (too much memory, big outputs from the endpoint), so we need to be cautious 
4. I was not aware of the knob for CQ when was leaving the comment https://github.com/kubernetes-sigs/kueue/pull/3673#discussion_r1861643242.

So, my proposal for 0.10 are:
1. keep as currently, just feature gate
2. add the knob just to enable / disable `EnableLocalQueueMetrics`

I think (2.) makes sense regardless if we do it for 0.10, so that we have a tool to control cardinality if it is a problem, and by analogy to the CQ knob - it does not make sense to the the knob for CQ and not for LQ. 

We can try to go ahead with this PR, and you can follow up with the knob (2.) - it either gets to 0.10 or not. WDYT?

For Beta, we can consider additional API with selectors, but I would refrain from doing so, until we have feedback indicating we really need them.

KPostOffice commented:
> We can try to go ahead with this PR as is (feature gate), and you can follow up with the knob (2.) - it either gets to 0.10 or not. WDYT?

This sounds good to me. I'll fix the comment you had and follow up this PR with the knob. Did you mean `EnableLocalQueueMetrics` or `EnabledLocalQueueResourceMetrics`?

I think cardinality would only be a concern for ResourceMetrics since that would be:

LocalQueues * Flavors * Resources

(or at least unique combinations of these)
All other metrics should be basically equal to # of LocalQueues

mimowo commented:
> This sounds good to me. I'll fix the comment you had and follow up this PR with the knob.

Cool
/lgtm
/approve

KPostOffice commented:
/retest-required

mimowo commented:
/retest
The integration test failed due to https://github.com/kubernetes-sigs/kueue/issues/3633

mimowo commented on cmd/kueue/main.go:
```diff
@@ -130,7 +130,9 @@ func main() {
 		os.Exit(1)
 	}
 
-	metrics.Register()
```
why is `metrics.Register()` gone? I would suggest to move`metrics.RegisterLQMetrics()` inside `metrics.Register()`

mimowo replied:
I think this was likely a bug which would cause no metrics at all, so I would like to have some e2e tests for metrics (separate PR)

mimowo reviewed (commented):
/lgtm
/approve
Please follow up with the knob to enable / disable. However, this is good to me as is for 0.10 since this is alpha (disabled by default) anyway.

mimowo commented:
/remove-kind api-change

mimowo commented:
/release-note-edit
```release-note
Introduce alpha feature, behind the LocalQueueMetrics feature gate, which allows users to get the prometheus LocalQueues metrics:
local_queue_pending_workloads
local_queue_quota_reserved_workloads_total
local_queue_quota_reserved_wait_time_seconds
local_queue_admitted_workloads_total
local_queue_admission_wait_time_seconds
local_queue_admission_checks_wait_time_seconds
local_queue_evicted_workloads_total
local_queue_reserving_active_workloads
local_queue_admitted_active_workloads
local_queue_status
local_queue_resource_reservation
local_queue_resource_usage
```
